### PR TITLE
update deployment backend to cephadm

### PR DESCRIPTION
**Pull request by batrick**
This ends up being significantly faster and ceph-ansible is on the way
out.

Obviously, this is not backwards compatible with Nautilus or older.

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
cc @kotreshhr

**Comment by batrick** [comment]
cc @bengland2

File: cephadm.yml [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      owner: root
+      content: "{{ cephadm.content }}"
+
+  - name: create /etc/ceph directory
+    file:
+      path: /etc/ceph
+      state: directory
+      owner: root
+      group: root
+      mode: 0755
```
Comment (sebastian-philipp) [inline cephadm.yml:33]: would be great, if we could move this upstream to cephadm: 

https://github.com/ceph/ceph/blob/980c286ce9b757985846261f2c377571fb1cd59e/src/cephadm/cephadm#L2416-L2417

File: cephadm.yml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    shell: /root/cephadm shell -- ceph config set global cluster_network 192.168.0.0/16
+
+  - name: log to files
+    shell: /root/cephadm shell -- ceph config set global log_to_file true
+
+  - name: log cluster to file
+    shell: /root/cephadm shell -- ceph config set global mon_cluster_log_to_file true
+
+  - name: disable stderr logging
+    shell: /root/cephadm shell -- ceph config set global log_to_stderr false
```
Comment (sebastian-philipp) [inline cephadm.yml:61]: interesting. Fyi, this is going to break `cephadm logs ...`.

File: cephadm.yml [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+- hosts: osds
+  become: yes
+  tasks:
+  - name: get hostname
+    shell: hostname --fqdn
+    register: myhostname
+
+#FIXME this is not idempotent!!
+  - name: add osd
+    shell: /root/cephadm shell -- ceph orch daemon add osd "{{ myhostname.stdout }}:/dev/sdc"
```
Comment (sebastian-philipp) [inline cephadm.yml:103]:
```bash
cat <<EOF | ceph orch apply -i 
service_type: osd
spec:
  placement:
    hosts:
    - {{ myhostname.stdout }}
  data_devices:
    paths: 
    - /dev/sdc
EOF
```

should be idempotent

File: cephadm.yml [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+
+- hosts: all
+  become: yes
+  tasks:
+  - name: generate min config
+    shell: /root/cephadm shell -- ceph config generate-minimal-conf
+    delegate_to: mon-000
+    register: minimal_config
+    run_once: true
```
Comment (sebastian-philipp) [inline cephadm.yml:131]: https://github.com/ceph/ceph/pull/35576

File: cephadm.yml (hunk 1 above)
Comment (batrick) [reply cephadm.yml:33]: You mean to create the `/etc/ceph` directory if it does not exist, yes that'd be good.

File: cephadm.yml (hunk 2 above)
Comment (batrick) [reply cephadm.yml]: That's fine I think? I have a set of scripts which collect logs from the usual `/var/log/ceph`.

File: cephadm.yml (hunk 3 above)
Comment (batrick) [reply cephadm.yml]:
````
fatal: [osd-010 -> None]: FAILED! => {
    "changed": true,
    "cmd": "tee /dev/stderr <<EOF | ceph orch apply -i -\nservice_type: osd\nspec:\n  placement:\n    hosts:\n    - li566-143.members.linode.com\n  data_devices:\n    paths:\n    - /dev/sdc\nEOF\n",
    "delta": "0:00:01.536072",
    "end": "2020-07-14 20:52:24.575292",
    "rc": 2,
    "start": "2020-07-14 20:52:23.039220"
}

STDERR:

service_type: osd
spec:
  placement:
    hosts:
    - li566-143.members.linode.com
  data_devices:
    paths:
    - /dev/sdc
Error ENOENT: Failed to validate Drive Group: Feature <spec> is not supported


MSG:

non-zero return code
````

?

**Comment by batrick** [comment]
Just leaving a note for posterity that this commit did not work:

````
commit 5a35a153d76999905f2f28092a07785f30d61968 HEAD (HEAD -> cephadm, origin/heads/cephadm)
gpg: Signature made Wed 15 Jul 2020 09:25:24 PM PDT
gpg:                using RSA key 88AD60FFD3FC04B6DB87CBAC3A2A7E25BEA8AADB
gpg: Good signature from "Patrick Joseph Donnelly (Hacker) <batrick@batbytes.com>" [ultimate]
gpg:                 aka "Patrick Joseph Donnelly (Software Engineer) <pdonnell@redhat.com>" [ultimate]
Author: Patrick Donnelly <pdonnell@redhat.com>
Date:   Tue Jul 14 13:26:09 2020 -0700

    testme
---
 cephadm.yml | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/cephadm.yml b/cephadm.yml
index c25b580..1b56053 100644
--- a/cephadm.yml
+++ b/cephadm.yml
@@ -133,9 +133,21 @@
     shell: hostname --fqdn
     register: myhostname
 
-#FIXME this is not idempotent!!
+#    shell: tee /dev/stderr | ceph orch apply -i
+#      stdin: |
   - name: add osd
-    shell: ceph orch daemon add osd "{{ myhostname.stdout }}:/dev/sdc"
+    shell:
+      cmd: |
+        tee /dev/stderr <<EOF | ceph orch apply -i -
+        service_type: osd
+        spec:
+          placement:
+            hosts:
+            - {{ myhostname.stdout }}
+          data_devices:
+            paths:
+            - /dev/sdc
+        EOF
     delegate_to: mon-000
     ignore_errors: yes
 

````

File: cephadm.yml (hunk 3 above)
Comment (sebastian-philipp) [reply cephadm.yml]: was backported just recently with https://github.com/ceph/ceph/pull/35537 . Maybe try again?

**Pull request merged**